### PR TITLE
[SHIRO-900] Refactor and update pom.xml

### PR DESCRIPTION
**Pull request by lprimak**
- Update Apache parent to 29 from 28
- Factor out jacoco properties and separate from maven surefire / failsafe argLine
- Update Groovy from Codehaus to Apache Groovy (4.x)
- Updade gmaven groovy plugin
- Update maven versions plugin
- Add compiler deprecation and unchecked lint by default
- Add JDK 11+ options for surefire and failsafe plugins

Following this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/SHIRO) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[SHIRO-XXX] - Fixes bug in SessionManager`,
       where you replace `SHIRO-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean install apache-rat:check` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If you have a group of commits related to the same change, please squash your commits into one and force push your branch using `git rebase -i`. 
 
Trivial changes like typos do not require a JIRA issue (javadoc, comments...). 
In this case, just format the pull request title like `(DOC) - Add javadoc in SessionManager`.
 
If this is your first contribution, you have to read the [Contribution Guidelines](https://github.com/apache/shiro/blob/master/CONTRIBUTING.md)

If your pull request is about ~20 lines of code you don't need to sign an [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) 
if you are unsure please ask on the developers list.

To make clear that you license your contribution under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -330,13 +333,15 @@
                             <exclude>**/spring.factories</exclude>
                             <exclude>**/spring.provides</exclude>
                             <exclude>**/*.iml</exclude>
+                            <exclude>**/nb-configuration.xml</exclude>
+                            <exclude>**/faces-config.NavData</exclude>
```
Comment (bmarwell) [inline pom.xml:337]: This could also be in the `.gitignore` file.

File: pom.xml [hunk 2]
```diff
@@ -543,6 +548,12 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.8.0</version>
+                <configuration>
+                    <compilerArgs>
+                        <arg>-Xlint:deprecation</arg>
+                        <arg>-Xlint:unchecked</arg>
+                    </compilerArgs>
+                </configuration>
```
Comment (bmarwell) [inline pom.xml:556]: what about maven.compiler.showWarnings (property) or showWarnings (configuration item)? Is it set, too?

File: pom.xml [hunk 3]
```diff
@@ -577,6 +588,9 @@
                                 <excludes>
                                     <exclude>**/main/**/samples/**</exclude>
                                 </excludes>
+                                <address>localhost</address>
+                                <port>${jacocoPort}</port>
+                                <propertyName>jacocoAgent</propertyName>
```
Comment (bmarwell) [inline pom.xml:593]: Why is this needed? Why was it not needed before?

File: pom.xml [hunk 4]
```diff
@@ -1462,6 +1474,8 @@
                         <exclude>**/spring.factories</exclude>
                         <exclude>**/spring.provides</exclude>
                         <exclude>**/*.iml</exclude>
+                        <exclude>**/nb-configuration.xml</exclude>
+                        <exclude>**/faces-config.NavData</exclude>
```
Comment (bmarwell) [inline pom.xml:1478]: Same as above

File: pom.xml [hunk 5]
```diff
@@ -1575,6 +1589,16 @@
...
             </build>
         </profile>
+        <profile>
+            <id>jdk11plus</id>
+            <activation>
+                <jdk>[11,)</jdk>
+            </activation>
+            <properties>
+                <failsafe.argLine>--add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.net=ALL-UNNAMED</failsafe.argLine>
+                <surefire.argLine>-Xshare:off ${failsafe.argLine}</surefire.argLine>
```
Comment (bmarwell) [inline pom.xml:1599]: -1, `-Xshare:off` will not work on all JDKs (like non-hotspot-JDKs as OpenJ9 which comes with IBM Semeru). Either additionally check for a hotspot JDK or just leave it out. It should not be required anyway.

File: pom.xml [hunk 6]
```diff
@@ -1575,6 +1589,16 @@
                 </pluginManagement>
             </build>
         </profile>
+        <profile>
+            <id>jdk11plus</id>
+            <activation>
+                <jdk>[11,)</jdk>
+            </activation>
+            <properties>
+                <failsafe.argLine>--add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.net=ALL-UNNAMED</failsafe.argLine>
```
Comment (bmarwell) [inline pom.xml:1598]: As far as I can see, this should not be needed anymore. Even then, we already have this in the guice projects so it should not live here for ALL modules. Only those which need it.

**Review by bmarwell — CHANGES_REQUESTED** [review]
Request for the profile (last 2 comments), rest is only comments/questions.

File: pom.xml (hunk 3 above)
Comment (lprimak) [reply pom.xml:593]: Currently, ITs do not participate in jacoco coverage. This is the infra to enable this (usage coming in the Jakarta EE PR)

File: pom.xml (hunk 5 above)
Comment (lprimak) [reply pom.xml:1599]: OpenJ9 (Semuru) works perfectly with `-Xshare:off`

File: pom.xml (hunk 1 above)
Comment (lprimak) [reply pom.xml]: removed

File: pom.xml (hunk 2 above)
Comment (lprimak) [reply pom.xml]: done

File: pom.xml (hunk 3 above)
Comment (lprimak) [reply pom.xml]: removed

File: pom.xml [hunk 7]
```diff
@@ -71,12 +71,14 @@
...
         <project.build.outputTimestamp>2022-03-22T23:08:15Z</project.build.outputTimestamp>
         <jacoco.skip>true</jacoco.skip>
+        <surefire.version>3.0.0-M7</surefire.version>
         <!--suppress CheckTagEmptyBody -->
         <surefire.argLine></surefire.argLine>
         <!--suppress CheckTagEmptyBody -->
         <failsafe.argLine></failsafe.argLine>
         <nexus.deploy.skip>false</nexus.deploy.skip>
-
+        <jacocoPort>52212</jacocoPort>
```
Comment (fpapon) [inline pom.xml:80]: Is it possible to use a random port? We can have some conflicts on CI platform.

Comment (bmarwell) [reply pom.xml:80]: Usually, it is not needed to specify a port. Even with integration tests. At least when using meecrowave, as it sits in the same JVM.

Comment (lprimak) [reply pom.xml:80]: Actually, this is no longer necessary

File: pom.xml [hunk 8]
```diff
@@ -1575,6 +1599,15 @@
...
         </profile>
+        <profile>
+            <id>jdk11plus</id>
+            <activation>
+                <jdk>[11,)</jdk>
+            </activation>
+            <properties>
+                <surefire.argLine>-XX:+IgnoreUnrecognizedVMOptions -Xshare:off</surefire.argLine>
+            </properties>
+        </profile>
```
Comment (bmarwell) [inline pom.xml:1610]: Small comment about why this exists would be appreciated.

File: pom.xml [hunk 9]
```diff
@@ -1462,6 +1485,7 @@
                         <exclude>**/spring.factories</exclude>
                         <exclude>**/spring.provides</exclude>
                         <exclude>**/*.iml</exclude>
+                        <exclude>**/target/**</exclude>
```
Comment (bmarwell) [inline pom.xml:1488]: Only informational: This should already be included due to `.gitignore` (but doesn't hurt here either).

**Review by bmarwell — APPROVED** [review]
LGTM, maybe add the comment but that doesn't prevent you from merging imho.

File: pom.xml (hunk 8 above)
Comment (lprimak) [reply pom.xml:1610]: done

File: pom.xml (hunk 9 above)
Comment (lprimak) [reply pom.xml:1488]: .gitignore wasn't working for target, because the rat plugin doesn't ignore what's underneath the directories and git does.

**Review by fpapon — APPROVED** [review]
@lprimak good job!
